<!-- 
    test_with_virtual_environment.launch
-->

<launch>
  <param name="use_sim_time" value="true"/>
<!-- Arguments Start -->
  <arg name="map"            default="auto_drive.bt"/>
  <arg name="bagfile"        default="autoDriveIntegration.bag"/>
  <arg name="playBagFile"    default="true"/>
  <arg name="snu_frame_id" default="SNU"/>
  <arg name="octomap_frame_id" default="map"/>
  <arg name="lane_width" default="6"/>

  <!--  Arguments for integration with KETI-->

  <arg name="vehicle_model/ly"            default="2.7"/> <!-- for the dynamics -->
  <arg name="vehicle_model/lx"            default="2"/> <!-- for safe margin -->
  <arg name="vehicle_model/steer_max"     default="0.52"/>
  <arg name="vehicle_model/accel_max"     default="1"/>
  <arg name="vehicle_model/accel_min"     default="-3"/>
  <arg name="vehicle_model/is_rear_wheel"     default="false"/>

  <arg name="nominal_speed"     default="2"/>
  <arg name="nominal_period"     default="5"/> <!--preirod = horizon -->
  <arg name="planning_dt"     default="0.1"/> <!--?-->

  <arg name="initial_goal/x" value="20" />

  <arg  name="initial_goal/y" value="68" />

  <!--Nominal radius-->
  <arg  name="obstacle_radius" value="0.8" />

<!-- Arguments End -->

<!-- Nodes Start -->
  <node pkg="octomap_server" type="octomap_server_node" name="octomap_server" output="screen" args="$(find atypical_driving)/worlds/$(arg map)" unless="$(arg playBagFile)">
    <param name="resolution"                 value="0.1" />
    <param name="frame_id"                   value="map" type="string"/>
    <param name="sensor_model/max_range"     value="-1" />
    <param name="height_map"                 value="false"/>
    <param name="color/r"                    value="0.2" />
    <param name="color/g"                    value="0.2" />
    <param name="color/b"                    value="0.2" />
    <param name="color/a"                    value="0.2" />

    <remap from="octomap_full" to="/atypical_planning_test/global_map"/>
  </node>

  <node pkg="rosbag" type="play" name="rosbag_record_diag" args="--clock --loop  $(find atypical_driving)/worlds/$(arg bagfile)" if="$(arg playBagFile)">
    <remap from="octomap_binary" to="/atypical_planning_test/local_map"/>
    <remap from="/airsim_car_node/object_pose" to="/atypical_planning_test/obstacle_pose"/>
  </node>

  <node pkg="atypical_driving" type="integration_server" name="atypical_planning_test" output="screen">

    <param name = "world_frame_id" value="map" />
    <param name = "snu_frame_id" value="$(arg snu_frame_id)" />
    <param name = "octomap_gen_frame_id" value="$(arg octomap_frame_id)" /> <!-- The frame id of octomap map or SNU-->

    <param name = "lane_csv_file" value = "$(find atypical_driving)/waypoint_airsim.csv"/>
    <param name = "lane_width" value="8"/>


    <param name="global_planner/horizon" value="$(arg nominal_period)" />
    <param name="global_planner/car_width" value="$(arg vehicle_model/lx)" />
    <param name="global_planner/car_z_min" value="0.5" />
    <param name="global_planner/car_z_max" value="1.0" />
    <param name="global_planner/car_speed" value="$(arg nominal_speed)" />
    <param name="global_planner/road_width" value="4.0" />
    <param name="global_planner/world_x_min" value="-10" />
    <param name="global_planner/world_y_min" value="-1" />
    <param name="global_planner/world_x_max" value="60" />
    <param name="global_planner/world_y_max" value="80" />
    <param name="global_planner/grid_resolution" value="0.5" />
    <param name="global_planner/box_resolution" value="0.3" />
    <param name="global_planner/box_max_size" value="10" />
    <param name="local_planner/horizon" value="$(arg nominal_period)" />
    <param name="local_planner/obstacle_radius_nominal" value="$(arg obstacle_radius)" />
    <param name="local_planner/ts" value="$(arg planning_dt)" />
    <param name="local_planner/car_longtitude" value="2.7" />
    <param name="local_planner/max_steer" value="0.52" />
    <param name="local_planner/max_accel" value="1.5" />
    <param name="local_planner/min_accel" value="-1.5" />

    <param name="initial_goal/x" value="$(arg initial_goal/x)" />
    <param name="initial_goal/y" value="$(arg initial_goal/y)" />

    <param name="predictor/observation_queue" value="4" />
    <param name="predictor/ref_height" value="0.4" />
  </node>

  <node pkg="atypical_driving" type="integration_client" name="atypical_driving_client" output="screen">
    <param name = "obstacle_radius" value="$(arg obstacle_radius)"/>
  </node>

  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find atypical_driving)/launch/rviz_config/config.rviz" output="log" />


<!-- Nodes End -->
</launch>
